### PR TITLE
fix discovery overwriting dynamic resources

### LINKED ISSUE #29828: Discovery Service overwrites dynamic resources of same name
Expected behavior:
- when a database or kube cluster is created or edited by a user and the discovery service discovers a database / kube cluster with the same name, it should not overwrite the dynamic resource that was manually configured by a user.

Current behavior:
- discovery service overwrites the resource if the name matches.

The reconciler attempts to check for this exact scenario here:
https://github.com/gravitational/teleport/blob/3163807a8a48bcbbf163494d92821a3666195a10/lib/services/reconciler.go#L158-L163

But the kube watcher and the database watcher both filter out existing resources that don't have `cloud` origin:
https://github.com/gravitational/teleport/blob/3163807a8a48bcbbf163494d92821a3666195a10/lib/srv/discovery/kube_watcher.go#L49
https://github.com/gravitational/teleport/blob/3163807a8a48bcbbf163494d92821a3666195a10/lib/srv/discovery/database_watcher.go#L99

As such, the "new" resource is not correctly identified as an already existing resource by the reconciler and skipped.

Bug details:
- Teleport version: master (pre-release v14)
- Recreation steps: create a dynamic database with the same name as a discovered database, and wait for the discovery service to overwrite it.

### PR DESCRIPTION
**Pull request by GavinFrazar**
Changelog: Prevent discovery service from overwriting Teleport dynamic resources that have the same name as discovered resources.

Closes https://github.com/gravitational/teleport/issues/29828

This fix relies on checking the existing resource to see if we can update it - when we add optimistic locking for db/kube_cluster/app then we can do a conditional update. It's very unlikely that a user will update the resource during such a small time window anyway.  I also consider using conditional updates out of scope to fix the linked issue because it's a fundamentally different problem.

### CODE REVIEW

**Review by rosstimothy — COMMENTED** [review]
How come we opted not to proceed with your [suggestion](https://github.com/gravitational/teleport/issues/29828#issuecomment-1659009706) from #29828? I think the Create, Get, Update sequence introduced here could pose other issues since none of the operations are atomic and we may be operating on outdated state by the time the Get on creation failure occurs.

**Comment by GavinFrazar** [comment]
> How come we opted not to proceed with your [suggestion](https://github.com/gravitational/teleport/issues/29828#issuecomment-1659009706) from #29828? I think the Create, Get, Update sequence introduced here could pose other issues since none of the operations are atomic and we may be operating on outdated state by the time the Get on creation failure occurs.

I opted to not do that because the reconciler would have to be changed to avoid deleting current resources in a different discovery_group or of a different origin - it seemed to me like that behavior was too specific to put in reconciler code and it would affect a ton of things that use the reconciler.

If I did do that then we would still be operating on potentially outdated state, right?
Reconciler could call `GetCurrentResources()`, then one of those resources is updated before it takes some action like create/delete/update.

File: lib/srv/discovery/database_watcher.go [hunk 1]
```diff
@@ -138,16 +138,17 @@ func (s *Server) getCurrentDatabases() map[string]types.Database {
 func (s *Server) onDatabaseCreate(ctx context.Context, database types.Database) error {
 	s.Log.Debugf("Creating database %s.", database.GetName())
 	err := s.AccessPoint.CreateDatabase(ctx, database)
-	// If the database already exists but has an empty discovery group, update it.
-	if trace.IsAlreadyExists(err) && s.updatesEmptyDiscoveryGroup(
-		func() (types.ResourceWithLabels, error) {
+	// If the database already exists but has cloud origin and an empty
+	// discovery group, then update it.
+	if err != nil {
+		err := s.resolveCreateErr(err, types.OriginCloud, func() (types.ResourceWithLabels, error) {
```
Comment (greedy52) [inline lib/srv/discovery/database_watcher.go:144]: Discovery service should only be able to get databases with `OriginCloud`:
https://github.com/gravitational/teleport/blob/89be51492373e53715f8ca24f6cca90f4b6852c9/lib/authz/permissions.go#L1188-L1191

Are you able to grab the old database that has a different origin? Based on the above permission, Update should/supposed fail too as Discovery service shouldn't have permission to the previous object for Update call.

Comment (GavinFrazar) [reply lib/srv/discovery/database_watcher.go:144]: I had db service running in the same process, maybe that's why it was allowed. Discovery needs to be able to Get any type of database. However I assumed that discovery service has this permission since it already calls GetDatabases for each reconciliation?

Comment (greedy52) [reply lib/srv/discovery/database_watcher.go:144]: My understanding is `GetDatabases` should only return the databases a user (or system user) has access to:
https://github.com/gravitational/teleport/blob/10932655f81c242cbbb56c76c29dae74486aff47/lib/auth/auth_with_roles.go#L5217-L5228

Related UT:
https://github.com/gravitational/teleport/blob/10932655f81c242cbbb56c76c29dae74486aff47/lib/auth/auth_with_roles_test.go#L2136-L2168

> I had db service running in the same process, maybe that's why it was allowed.

Which service has full RW on `types.KindDatabase`? Are you running Discovery with Auth? So per API call we can't tell which role it is calling with? that's tough

Comment (GavinFrazar) [reply lib/srv/discovery/database_watcher.go:144]: We run discovery with auth service in cloud:
```
$ k exec -it deploy/teleport-auth -- tctl inventory status --connected                   (dev) 
Server ID                            Services       Version Upgrader 
------------------------------------ -------------- ------- -------- 
9d88ebf9-89ba-4076-99e1-f56f74ab235c Proxy          16.0.4  none     
3158eaa8-699f-4c5f-ab52-ec7e27fa8e75 Node           15.4.4  unit     
af09e479-3ecb-4295-bf1f-3745ded15f4c Auth,Discovery 16.0.4  none    
```

However, I could not reproduce the overwrite bug with my cloud tenant.

I am able to reproduce the bug if I run discovery service locally. I have several other services in there:
ssh, kube, db, app, discovery all in one config.

I think what's happening is that our "verb rules" check doesn't apply resource labels.
We just check if the identity is allowed to use VerbUpdate:
- https://github.com/gravitational/teleport/blob/f1dd572a2b55425ac1ecb7e3e601cb1289144bde/lib/auth/auth_with_roles.go#L6335-L6349

After that check passes, we check if they have "access" to the database, but the "access" check doesn't check the verb.

So basically, it's allowed to update the resource because `discovery_service` has the "update" verb allowed, and it's allowed to "access" the database because some other role like `db_service` has labels that match the database.

I'm not sure why it doesn't happen on cloud, but I suspect its due to the fact that auth doesn't use (nor need) a role with database labels in order to access databases.

I think for now at least, we have to assume there are some cases where discovery has permission to update the databases. It's a limitation of our RBAC.

**Review by greedy52 — APPROVED** [review]

File: lib/srv/discovery/discovery.go [hunk 2]
```diff
@@ -1686,14 +1686,37 @@ func splitMatchers[T types.Matcher](matchers []T, matcherTypeCheck func(string)
...
-		return false
+		return trace.NewAggregate(createErr, err)
+	}
+
+	// Check that the registered resource origin matches the origin we want.
+	oldOrigin, err := types.GetOrigin(old)
+	if err != nil {
+		return trace.NewAggregate(createErr, err)
+	}
+	if oldOrigin != discoveryOrigin {
```
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/srv/discovery/discovery.go:1706]: We check if the old group is not empty below because that's a special case where the old/new groups don't match but we still want to update the resource.
In this way, discovery agents with a `discovery_group` essentially claim the resources they discover that used to be (or currently are) discovered by an agent that did not have the group configured.

For origin, we don't want that behavior - empty origin should not be updated because it's definitely not managed by discovery service.

[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/srv/discovery/discovery.go:1706]: since anyone reading this code might make the same assumption (admittedly the intent here is subtle), I went ahead and added test cases for empty origin and a comment explaining the above as well.

[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/srv/discovery/discovery.go:1706]: https://github.com/gravitational/teleport/compare/f1dd572a2b55425ac1ecb7e3e601cb1289144bde..1190fed6c6047125c3e0518736ef6c589086c9be

[the remark this answers is not in the document]
Comment (greedy52) [reply lib/srv/discovery/discovery.go:1706]: Databases should always have an origin:
https://github.com/gravitational/teleport/blob/ca0c0f4fd9e844c1bf11105afe1c284bf78cf911/lib/auth/grpcserver.go#L3715-L3722

**Review by rosstimothy — APPROVED** [review]

**Comment by milos-teleport** [comment]
Example error if someone stumbles upon this with v14 or earlier

```
2024-01-01T15:20:00Z WARN [DISCOVERY] Unable to reconcile database resources. error:[
ERROR REPORT:
Original Error: trace.aggregate failed to create db dbname
database &#34;dbname&#34; doesn&#39;t exist
```
